### PR TITLE
Fix for issue #800.

### PR DESCRIPTION
**Pull request by romaxa**
Add custom style diffing which perform better than { diff: true } with default mapbox API

### CODE REVIEW

File: src/components/static-map.js [hunk 1]
```diff
@@ -197,14 +206,56 @@ export default class StaticMap extends PureComponent<StaticMapProps, State> {
     }
   }
 
+  _tryToApplyLatestMapChanges() {
```
Comment (ibgreen) [inline src/components/static-map.js]: This looks subtle. If this was "copy-pasted", add link to the source, if not add comments to guide maintainers?

Comment (vicapow) [reply src/components/static-map.js]: this was copied from an internal project. once this is in react-map-gl, we'll replace our internal implementation with this one.

File: src/components/static-map.js [hunk 2]
```diff
@@ -31,6 +32,7 @@ import mapboxgl from '../utils/mapboxgl';
 import {checkVisibilityConstraints} from '../utils/map-constraints';
 import {MAPBOX_LIMITS} from '../utils/map-state';
 import MapContext from './map-context';
+import {diff as diffStyles} from '@mapbox/mapbox-gl-style-spec';
```
Comment (Pessimistress) [inline src/components/static-map.js]: This package is published from https://github.com/mapbox/mapbox-gl-js/tree/master/src/style-spec and is exactly what mapbox uses to do diffing:
https://github.com/mapbox/mapbox-gl-js/blob/master/src/style/style.js#L30
https://github.com/mapbox/mapbox-gl-js/blob/master/src/style/style.js#L435

If this issue is caused by a bug in mapbox-gl it should be fixed in their repo.

File: src/components/static-map.js (hunk 1 above)
Comment (Pessimistress) [reply src/components/static-map.js]: We made an explicit effort to move away from custom diffing in v3. What property in your style is causing the failure, shouldn't we open an issue on mapbox-gl?

File: src/components/static-map.js (hunk 2 above)
Comment (vicapow) [reply src/components/static-map.js]: I believe this is a difference in behavior between the two libraries. mapbox-gl chooses to error if it cannot update the style just yet. In our case, we want to update the map for the user when we can, retrying if needed.

Comment (Pessimistress) [reply src/components/static-map.js]: If the error is just caused by map not being loaded, why do we need to implement our own diffing logic? Can we just call `map.setStyle` when it's ready?

Comment (vicapow) [reply src/components/static-map.js]: hrm. good point

File: src/components/static-map.js (hunk 1 above)
Comment (romaxa) [reply src/components/static-map.js]: is-stroke - is failing to diff
args: [
0: "layerId"
1: "is-stroke"
2: true
]
command: "setLayerProperty"

File: src/components/static-map.js (hunk 2 above)
Comment (romaxa) [reply src/components/static-map.js]: I've looked at [this part of code](https://github.com/mapbox/mapbox-gl-js/blob/master/src/style/style.js#L435) with debugger and found that 
this.serialize() is very different to what is nextState, using simple diff tool.

It generates thousands of changes for exactly same style layers every time.
Do we need to normalize nextState somehow before diffing?

In this code change, we are diffing styles without mutations, so it only detects real changes

**Comment by Pessimistress** [comment]
react-map-gl is no longer performing style diffing itself.

**Pull request closed without merging**